internal/compiler-bridge/src-2.10/main/scala/xsbt/ExtractUsedNames.scala

```diff
@@ -57,57 +84,81 @@ class ExtractUsedNames[GlobalType <: CallbackGlobal](val global: GlobalType) ext
-            ()
-          }
-          selectors foreach { selector =>
-            usedNameInImportSelector(selector.name)
-            usedNameInImportSelector(selector.rename)
-          }
-        // TODO: figure out whether we should process the original tree or walk the type
-        // the argument for processing the original tree: we process what user wrote
-        // the argument for processing the type: we catch all transformations that typer applies
-        // to types but that might be a bad thing because it might expand aliases eagerly which
-        // not what we need
-        case t: TypeTree if t.original != null =>
-          t.original.foreach(handleTreeNode)
-        case t if t.hasSymbol && eligibleAsUsedName(t.symbol) =>
-          addSymbol(t.symbol)
-        case _ => ()
+    private def addName(name: Name, enclosingNonLocalClass: Symbol = resolveEnclosingNonLocalClass): Unit = {
+      val nameAsString = name.decode.trim
+      if (enclosingNonLocalClass == NoSymbol || enclosingNonLocalClass.isPackage) {
+        namesUsedAtTopLevel += nameAsString
+      } else {
+        val className = ExtractUsedNames.this.className(enclosingNonLocalClass)
+        namesUsedInClasses(className) += nameAsString
       }
+    }
+
+    private def handleMacroExpansion(original: Tree): Unit = {
+      original.foreach(traverse)
+    }
 
-      node match {
-        case MacroExpansionOf(original) if inspectedOriginalTrees.add(original) =>
-          handleClassicTreeNode(node)
-          handleMacroExpansion(original)
-        case _ =>
-          handleClassicTreeNode(node)
+    private def handleClassicTreeNode(tree: Tree): Unit = tree match {
+      case _: DefTree | _: Template => ()
+      // turns out that Import node has a TermSymbol associated with it
+      // I (Grzegorz) tried to understand why it's there and what does it represent but
+      // that logic was introduced in 2005 without any justification I'll just ignore the
+      // import node altogether and just process the selectors in the import node
+      case Import(_, selectors: List[ImportSelector]) =>
+        val enclosingNonLocalClass = resolveEnclosingNonLocalClass
+        def usedNameInImportSelector(name: Name): Unit =
+          if ((name != null) && (name != nme.WILDCARD)) addName(name, enclosingNonLocalClass)
+        selectors foreach { selector =>
+          usedNameInImportSelector(selector.name)
+          usedNameInImportSelector(selector.rename)
+        }
+      // TODO: figure out whether we should process the original tree or walk the type
+      // the argument for processing the original tree: we process what user wrote
+      // the argument for processing the type: we catch all transformations that typer applies
+      // to types but that might be a bad thing because it might expand aliases eagerly which
+      // not what we need
+      case t: TypeTree if t.original != null =>
+        t.original.foreach(traverse)
```

Comment:
I believe this line (and the one above for macro originals) is a performance regression. Both `foreach` and `traverse` will recurse into the tree. See https://github.com/sbt/zinc/pull/187

Either `t.original.foreach(handleTreeNode)` or simply `traverse(t.original)` seems to be the right approach.